sdk/src/logs/batch_log_processor.cc

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "opentelemetry/sdk/logs/batch_log_processor.h"
+
+#include <vector>
+using opentelemetry::sdk::common::AtomicUniquePtr;
+using opentelemetry::sdk::common::CircularBufferRange;
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace logs
+{
+BatchLogProcessor::BatchLogProcessor(std::unique_ptr<LogExporter> &&exporter,
+                                     const size_t max_queue_size,
+                                     const std::chrono::milliseconds scheduled_delay_millis,
+                                     const size_t max_export_batch_size)
+    : exporter_(std::move(exporter)),
+      max_queue_size_(max_queue_size),
+      scheduled_delay_millis_(scheduled_delay_millis),
+      max_export_batch_size_(max_export_batch_size),
+      buffer_(max_queue_size_),
+      worker_thread_(&BatchLogProcessor::DoBackgroundWork, this)
+{}
+
+std::unique_ptr<Recordable> BatchLogProcessor::MakeRecordable() noexcept
+{
+  return exporter_->MakeRecordable();
+}
+
+void BatchLogProcessor::OnReceive(std::unique_ptr<Recordable> &&record) noexcept
+{
+  if (is_shutdown_.load() == true)
+  {
+    return;
+  }
+
+  if (buffer_.Add(record) == false)
+  {
+    return;
+  }
+
+  // If the queue gets at least half full a preemptive notification is
+  // sent to the worker thread to start a new export cycle.
+  if (buffer_.size() >= max_queue_size_ / 2)
+  {
+    // signal the worker thread
+    cv_.notify_one();
+  }
+}
+
+bool BatchLogProcessor::ForceFlush(std::chrono::microseconds timeout) noexcept
+{
+  if (is_shutdown_.load() == true)
+  {
+    return false;
+  }
+
+  is_force_flush_ = true;
+
+  // Keep attempting to wake up the worker thread
```

Comment:
Could you elaborate on why the worker may not immediately change the state. We are effectively burning CPU cycles here..?

Reply:
Yeah I think we're in agreement - Tom also mentioned this in an earlier comment.

I had removed it in an earlier commit [here](https://github.com/open-telemetry/opentelemetry-cpp/pull/434/commits/14033b514096017b57d9c8cd00217e383b2798e4#diff-f7c189661e15834743b64c453ad9e3548337c19515074da4bc43bb0205c94ffaR179) and explained why [in this comment](https://github.com/open-telemetry/opentelemetry-cpp/pull/434#discussion_r543525388) but didn’t receive any replies so i wasnt sure if it was good or not